src/iterator.cr

```diff
@@ -131,13 +131,19 @@ module Iterator(T)
   end
 
   def self.of(&block : -> T)
-    SingletonProc(T).new(block)
+    SingletonProc(typeof(without_stop(&block))).new(block)
+  end
+
+  private def self.without_stop(&block : -> T)
+    e = block.call
+    raise "" if e.is_a?(Iterator::Stop)
```

Comment:
We could find probably a more idiomatic way to express this. maybe a `fail : NoReturn` or `halt` or `unreachable` that will abort compilation at codegen but will return `NoReturn` in semantic could work better. #someday